ansible/roles/test/files/ptftests/inner_hash_test.py

```diff
@@ -0,0 +1,505 @@
+
+        pkt = self.generate_inner_pkt(sport, dport, ip_src, ip_dst, ip_proto)
+        nvgre_pkt = simple_nvgre_packet(
+                    eth_dst=self.router_mac,
+                    eth_src=src_mac,
+                    ip_id=0,
+                    ip_src=outer_ip_src,
+                    ip_dst=outer_ip_dst,
+                    ip_ttl=64,
+                    nvgre_tni=random.randint(1, 254)+20000,
+                    inner_frame=pkt)
+
+        logging.info("Sending packet from port {} outer_ip_src {} outer_ip_dst {} inner_src_ip {} inner_dst_ip {} inner_sport {} inner_dport {} inner_ipproto {}"\
+                .format(src_port, outer_ip_src, outer_ip_dst, ip_src, ip_dst, sport, dport, ip_proto))
+        send_packet(self, src_port, nvgre_pkt)
+
+        masked_exp_pkt = Mask(nvgre_pkt)
+        masked_exp_pkt.set_do_not_care_scapy(scapy.Ether, "dst")
+        masked_exp_pkt.set_do_not_care_scapy(scapy.Ether, "src")
+        masked_exp_pkt.set_do_not_care_scapy(scapy.IP, "chksum")
+        masked_exp_pkt.set_do_not_care_scapy(scapy.IP, "ttl")
+
+        return verify_packet_any_port(self, masked_exp_pkt, self.exp_port_list)
+
+
+    def simple_nvgrev6_packet(self, pktlen=300,
+                        eth_dst='00:01:02:03:04:05',
+                        eth_src='00:06:07:08:09:0a',
+                        dl_vlan_enable=False,
+                        vlan_vid=0,
+                        vlan_pcp=0,
+                        dl_vlan_cfi=0,
+                        ipv6_src='1::2',
+                        ipv6_dst='3::4',
+                        ipv6_fl=0,
+                        ipv6_tc=0,
+                        ipv6_ecn=None,
+                        ipv6_dscp=None,
+                        ipv6_hlim=64,
+                        nvgre_version=0,
+                        nvgre_tni=None,
+                        nvgre_flowid=0,
+                        inner_frame=None
+                        ):
+        '''
+        @summary: Helper function to construct an IPv6 NVGRE packet
+        '''
+        if scapy.NVGRE is None:
+            logging.error("A NVGRE packet was requested but NVGRE is not supported by your Scapy. See README for more information")
+            return None
+
+        if MINSIZE > pktlen:
+            pktlen = MINSIZE
+
+        nvgre_hdr = scapy.NVGRE(vsid=nvgre_tni, flowid=nvgre_flowid)
+
+        #ip_tos = ip_make_tos(ip_tos, ip_ecn, ip_dscp)
+
+        if (dl_vlan_enable):
+            pkt = scapy.Ether(dst=eth_dst, src=eth_src)/ \
+                scapy.Dot1Q(prio=vlan_pcp, id=dl_vlan_cfi, vlan=vlan_vid)/ \
+                scapy.IPv6(src=ipv6_src, dst=ipv6_dst, fl=ipv6_fl, tc=ipv6_tc, hlim=ipv6_hlim, nh=47)/ \
+                nvgre_hdr
+        else:
+            pkt = scapy.Ether(dst=eth_dst, src=eth_src)/ \
+                scapy.IPv6(src=ipv6_src, dst=ipv6_dst, fl=ipv6_fl, tc=ipv6_tc, hlim=ipv6_hlim, nh=47)/ \
```

Comment:
@anish-n please don't use a magic number for GRE proto

Reply:
This is a well known number to identify GRE, since its a test, I don't consider it problematic. This is how the PTF framework themselves implement their testutils, it pretty similar to that